File: app/models/product/inventory.rb
# == Schema Information
#
# Table name: product_inventories
#
#  id                      :bigint           not null, primary key
#  location_id             :bigint           not null
#  unit                    :enum             default("pcs"), not null
#  quantity_on_hand        :float            default(0.0), not null
#  quantity_reserved       :float            default(0.0), not null
#  low_inventory_threshold :float            default(0.0), not null
#  expires_on              :date
#  created_at              :datetime         not null
#  updated_at              :datetime         not null
#
class Product::Inventory < ApplicationRecord

  # Included Modules:
  include UnitsOfMeasure

  # Virtual Attributes:
  attr_accessor :temporary_id

  # Enumerations:
  enum unit: UNITS_OF_MEASURE

  # Associations:
  belongs_to :location, inverse_of: :inventories
  has_many :variant_inventories, class_name: "Product::VariantInventory", foreign_key: "product_inventory_id", dependent: :destroy, inverse_of: :inventory
  has_many :variants, class_name: "Product::Variant", through: :variant_inventories

  # Validations:
  validates :quantity_on_hand, :quantity_reserved, :low_inventory_threshold, numericality: { greater_than_or_equal_to: 0 }
  validates :unit, inclusion: { in: UNITS_OF_MEASURE.values }
end

File: app/assets_webpack/src/javascripts/global/base/productForm/inventory.ts
import Unit from "../unit"
import { v4 as uuidv4 } from 'uuid'

export default class Inventory {
  temporaryId: string
  locationId: number
  locationName: KnockoutObservable<string>
  quantityOnHand: KnockoutObservable<number>
  lowInventoryThreshold: KnockoutObservable<number>
  expiresIn: KnockoutObservable<Date>
  unit: KnockoutObservable<Unit>

  constructor(params: {
    locationId: number,
    locationName: string,
    quantityOnHand?: number,
    lowInventoryThreshold?: number,
    unit?: { 
      name: string,
      measurementType: string
    }
  }) {
    this.temporaryId           = uuidv4()
    this.locationId            = params.locationId
    this.locationName          = ko.observable(params.locationName)
    this.quantityOnHand        = ko.observable(params.quantityOnHand || 1)
    this.lowInventoryThreshold = ko.observable(params.lowInventoryThreshold || 1)
    this.unit                  = ko.observable()
  }

  get serialize() {
    return {
      temporary_id: this.temporaryId,
      quantity_on_hand: this.quantityOnHand(),
      low_inventory_threshold: this.lowInventoryThreshold(),
      location_id: this.locationId,
      unit: this.unit().name(),
      expires_on: null
    }
  }

  get valid() {
    return !!this.quantityOnHand() &&
           !!this.lowInventoryThreshold() &&
           !!this.unit()
  }

  init() {
    return this
  }
}